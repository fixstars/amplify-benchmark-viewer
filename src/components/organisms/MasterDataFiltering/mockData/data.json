[
  { "name": "Tsp" },
  { "name": "Qap" },
  { "name": "MaxCut" },
  { "name": "Sudoku" },
  { "name": "Cvrp" }
]
