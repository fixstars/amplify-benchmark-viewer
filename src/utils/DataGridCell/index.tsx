/**
 * Copyright (c) Fixstars Corporation and Fixstars Amplify Corporation.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { LinkButton } from 'components/atoms/LinkButton'

const clientColors = [
  '#0f66dc',
  '#f57bae',
  '#e75e75',
  '#e5f97c',
  '#1b000d',
  '#f6a695',
  '#692981',
  '#3cde29',
  '#591414',
  '#d66594',
  '#2d884a',
  '#400763',
  '#9d19f4',
  '#fcdeb9',
  '#1f4c48',
  '#909743',
  '#2dc1c8',
  '#c17913',
  '#b2a1f6',
]

const instanceColors = [
  '#ff66ff',
  '#0f66dc',
  '#f57bae',
  '#e75e75',
  '#e5f97c',
  '#1b000d',
  '#f6a695',
  '#692981',
  '#3cde29',
  '#591414',
  '#d66594',
  '#2d884a',
  '#400763',
  '#9d19f4',
  '#fcdeb9',
  '#718de6',
  '#1f4c48',
  '#909743',
  '#2dc1c8',
  '#c17913',
  '#b2a1f6',
  '#9a916d',
  '#406a0f',
  '#004b31',
  '#18729f',
  '#9997c3',
  '#9cf0ec',
  '#ce843c',
  '#2d326e',
  '#76ed60',
  '#d7f7e8',
  '#bb15d7',
  '#87a7d8',
  '#94053a',
  '#e5c8a6',
  '#e2f2c0',
]

export interface ProblemInstances {
  readonly [key: string]: ReadonlyArray<string> | undefined
}

const renderClientsCell = (key: string, clients: ReadonlyArray<string>) => {
  const getBackgroundColor = (client: string): string => {
    const cache = JSON.parse(localStorage.getItem('clientColorCache') ?? '{}')
    let cacheIndex = JSON.parse(
      localStorage.getItem('clientColorIndexCache') ?? '-1',
    )
    let backgroundColor = cache[client]

    if (backgroundColor === undefined) {
      const length = clientColors.length
      cacheIndex = cacheIndex >= length ? 0 : cacheIndex + 1
      backgroundColor = clientColors[cacheIndex]
      cache[client] = backgroundColor
      localStorage.setItem('clientColorCache', JSON.stringify(cache))
      localStorage.setItem('clientColorIndexCache', JSON.stringify(cacheIndex))
      return backgroundColor
    }

    return backgroundColor
  }

  return clients.map((client) => (
    <div key={`${key}_${client}`} style={{ marginRight: 10 }}>
      <LinkButton
        label={client}
        link={`/clients/${client}`}
        backgroundColor={getBackgroundColor(client)}
      />
    </div>
  ))
}

const renderProblemInstancesCell = (
  key: string,
  problemInstances: ProblemInstances,
) => {
  const getBackgroundColor = (instance: string): string => {
    const cache = JSON.parse(localStorage.getItem('instanceColorCache') ?? '{}')
    let cacheIndex = JSON.parse(
      localStorage.getItem('instanceColorIndexCache') ?? '-1',
    )
    let backgroundColor = cache[instance]

    if (backgroundColor === undefined) {
      const length = instanceColors.length
      cacheIndex = cacheIndex >= length ? 0 : cacheIndex + 1
      backgroundColor = instanceColors[cacheIndex]
      cache[instance] = backgroundColor
      localStorage.setItem('instanceColorCache', JSON.stringify(cache))
      localStorage.setItem(
        'instanceColorIndexCache',
        JSON.stringify(cacheIndex),
      )
      return backgroundColor
    }

    return backgroundColor
  }
  const pairs = Object.entries(problemInstances)
  pairs.sort(function (p1, p2) {
    const p1Key = p1[0]
    const p2Key = p2[0]
    if (p1Key < p2Key) {
      return -1
    }
    if (p1Key > p2Key) {
      return 1
    }
    return 0
  })
  problemInstances = Object.fromEntries(pairs)
  return Object.keys(problemInstances).map((className) => (
    <div
      key={`${key}_${className}`}
      style={{ margin: '5px 0 5px 0', display: 'flex' }}
    >
      <span style={{ marginRight: 10 }}>{className}:</span>
      <div style={{ display: 'flex', flexWrap: 'wrap' }}>
        {problemInstances[className]?.map((instance) => (
          <div
            key={`${key}_${className}_${instance}`}
            style={{ marginRight: 10, marginBottom: 10 }}
          >
            <LinkButton
              label={instance}
              link={`/classes/${className}/instances/${instance}`}
              backgroundColor={getBackgroundColor(instance)}
            />
          </div>
        ))}
      </div>
    </div>
  ))
}

export { renderClientsCell, renderProblemInstancesCell }
