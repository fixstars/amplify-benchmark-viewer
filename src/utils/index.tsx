export * from './DataGridCell'
export * from './statistics'
