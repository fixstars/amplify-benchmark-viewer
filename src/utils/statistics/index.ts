function quartile(data: readonly number[], q: number) {
  const sortedData = data.slice().sort((a, b) => a - b)
  const pos = Math.floor((sortedData.length - 1) * q)
  return sortedData[pos]
}
function quartile0(data: readonly number[]) {
  return quartile(data, 0)
}
function quartile25(data: readonly number[]) {
  return quartile(data, 0.25)
}

function quartile50(data: readonly number[]) {
  return quartile(data, 0.5)
}

function quartile75(data: readonly number[]) {
  return quartile(data, 0.75)
}

function quartile100(data: readonly number[]) {
  return quartile(data, 1)
}

export { quartile0, quartile25, quartile50, quartile75, quartile100 }
