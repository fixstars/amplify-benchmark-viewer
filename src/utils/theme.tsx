/**
 * Copyright (c) Fixstars Corporation and Fixstars Amplify Corporation.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { createTheme } from '@mui/material/styles'

export const theme = createTheme({
  palette: {
    primary: {
      main: '#111',
    },
  },
  typography: {
    h2: {
      fontSize: '1.5rem',
    },
  },
  components: {
    MuiTooltip: {
      styleOverrides: {
        tooltip: {
          backgroundColor: '#111',
          color: '#f4f1f4',
          font: '10px',
        },
      },
    },
  },
})
