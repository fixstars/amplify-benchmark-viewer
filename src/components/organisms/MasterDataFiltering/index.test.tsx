/**
 * Copyright (c) Fixstars Corporation and Fixstars Amplify Corporation.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { render, fireEvent, screen, act } from '@testing-library/react'
import { mockBox } from 'utils/test'

import { MasterDataFiltering } from '.'
import mockData from './mockData/data.json'

const mockFilteringTable = jest.fn()

jest.mock('../FilteringTable', () => {
  const { FilteringTable: FilteringTableComponent, ...rest } =
    jest.requireActual('../FilteringTable')

  const FilteringTable = (props: typeof FilteringTableComponent) => {
    mockFilteringTable(props)
    return <FilteringTableComponent {...props} />
  }

  return {
    FilteringTable,
    ...rest,
  }
})

describe('<MasterDataFiltering />', () => {
  it('Rendered well', async () => {
    const { container } = render(
      <MasterDataFiltering
        label="Problem Class Filtering"
        data={mockData}
        selectedData={[]}
        onSelected={jest.fn}
      />,
    )

    let box = mockBox.mock.calls[0][0]
    expect(box.sx).toEqual({ marginBottom: '4px' })

    box = box.children
    expect(box.type.name).toBe('Box')
    expect(box.props.sx).toEqual({ display: 'flex', alignItems: 'flex-start' })

    const contents = box.props.children
    expect(contents.length).toBe(2)

    box = contents[0]
    expect(box.type.name).toBe('Box')
    expect(box.props.sx).toEqual({
      display: 'flex',
      alignItems: 'center',
      minWidth: '140px',
    })
    expect(box.props.children.length).toBe(3)

    {
      const label = box.props.children[0]
      expect(label.type.name).toBe('Typography')
      expect(label.props.children).toBe('Problem Class Filtering')
    }

    {
      const addIcon = box.props.children[1]
      expect(addIcon.type.render.name).toBe('IconButton')
      expect(addIcon.props.size).toBe('small')
      expect(addIcon.props.onClick.name).toBe('onClick')
    }

    {
      const menu = box.props.children[2]
      expect(menu.type.render.name).toBe('Menu')
      expect(menu.props.anchorEl).toBe(null)
      expect(menu.props.open).toBe(false)
      expect(menu.props.onClose.name).toBe('onClose')
      expect(menu.props.sx).toEqual({
        'li:hover': {
          backgroundColor: '#FFF',
        },
      })

      const filteringContainer = menu.props.children
      expect(filteringContainer.type.name).toBe('Box')
      expect(filteringContainer.props.sx).toEqual({
        width: '500px',
        height: '400px',
        padding: '4px 8px',
      })

      const filter = filteringContainer.props.children
      expect(filter.type.name).toBe('FilteringTable')
      expect(filter.props.title).toBe('Problem Class Filtering')
      expect(filter.props.selectedData).toEqual([])
      expect(filter.props.data).toEqual(mockData)
    }

    box = contents[1]
    expect(box.type.name).toBe('Box')

    const chips = box.props.children
    expect(chips.length).toBe(0)

    expect(container).toMatchSnapshot()
  })

  it('Rendered well with selected data', async () => {
    const { container } = render(
      <MasterDataFiltering
        label="Problem Class Filtering"
        data={mockData}
        selectedData={['Tsp', 'Sudoku']}
        onSelected={jest.fn}
      />,
    )

    let box = mockBox.mock.calls[0][0]
    box = box.children

    const contents = box.props.children
    box = contents[0]

    {
      const menu = box.props.children[2]
      const filteringContainer = menu.props.children
      const filter = filteringContainer.props.children
      expect(filter.props.selectedData).toEqual(['Tsp', 'Sudoku'])
    }

    box = contents[1]
    expect(box.type.name).toBe('Box')

    const chips = box.props.children
    expect(chips.length).toBe(2)

    {
      const chip = chips[0]
      expect(chip.key).toBe('master_data_filtering_chip_Tsp')
      expect(chip.type.render.name).toBe('Chip')
      expect(chip.props.label).toBe('Tsp')
      expect(chip.props.variant).toBe('outlined')
      expect(chip.props.size).toBe('small')
      expect(chip.props.sx).toEqual({ marginRight: '8px', marginBottom: '4px' })
      expect(chip.props.color).toBe('primary')
      expect(chip.props.onDelete.name).toBe('onDelete')
    }
    {
      const chip = chips[1]
      expect(chip.key).toBe('master_data_filtering_chip_Sudoku')
      expect(chip.type.render.name).toBe('Chip')
      expect(chip.props.label).toBe('Sudoku')
      expect(chip.props.variant).toBe('outlined')
      expect(chip.props.size).toBe('small')
      expect(chip.props.sx).toEqual({ marginRight: '8px', marginBottom: '4px' })
      expect(chip.props.color).toBe('primary')
      expect(chip.props.onDelete.name).toBe('onDelete')
    }

    expect(container).toMatchSnapshot()
  })

  it('Show filtering table', async () => {
    render(
      <MasterDataFiltering
        label="Problem Class Filtering"
        data={mockData}
        selectedData={[]}
        onSelected={jest.fn}
      />,
    )

    let box = mockBox.mock.calls[0][0]
    box = box.children

    let contents = box.props.children
    box = contents[0]

    {
      const addIcon = box.props.children[1]
      expect(addIcon.type.render.name).toBe('IconButton')
      expect(addIcon.props.size).toBe('small')
      expect(addIcon.props.onClick.name).toBe('onClick')

      const menu = box.props.children[2]
      expect(menu.props.anchorEl).toBe(null)
      expect(menu.props.open).toBe(false)

      // click add icon
      mockBox.mockClear()
      fireEvent.click(screen.getByTestId('AddCircleIcon'))
    }

    box = mockBox.mock.calls[0][0]
    box = box.children

    contents = box.props.children
    box = contents[0]

    {
      const menu = box.props.children[2]
      expect(menu.props.anchorEl.type).toBe('button')
      expect(menu.props.open).toBe(true)
    }
  })

  describe('onSelected', () => {
    it('add selected items via filtering table', async () => {
      const onMockSelected = jest.fn()
      render(
        <MasterDataFiltering
          label="Problem Class Filtering"
          data={mockData}
          selectedData={[]}
          onSelected={onMockSelected}
        />,
      )

      mockBox.mockClear()
      fireEvent.click(screen.getByTestId('AddCircleIcon'))

      expect(onMockSelected.mock.calls.length).toBe(0)

      act(() => {
        mockFilteringTable.mock.calls[0][0].onSelected(['Tsp', 'Sudoku'])
      })

      expect(onMockSelected.mock.calls[0][0]).toEqual(['Tsp', 'Sudoku'])
    })

    it('delete selected item via chips', async () => {
      const onMockSelected = jest.fn()

      render(
        <MasterDataFiltering
          label="Problem Class Filtering"
          data={mockData}
          selectedData={['Tsp', 'Sudoku']}
          onSelected={onMockSelected}
        />,
      )

      let box = mockBox.mock.calls[0][0]
      box = box.children

      const contents = box.props.children
      box = contents[0]

      {
        const menu = box.props.children[2]
        const filteringContainer = menu.props.children
        const filter = filteringContainer.props.children
        expect(filter.props.selectedData).toEqual(['Tsp', 'Sudoku'])
      }

      box = contents[1]
      expect(box.type.name).toBe('Box')

      const chips = box.props.children
      expect(chips.length).toBe(2)

      {
        const chip = chips[1]
        expect(chip.props.label).toBe('Sudoku')
        expect(chip.props.onDelete.name).toBe('onDelete')

        expect(onMockSelected.mock.calls.length).toBe(0)
        act(() => chip.props.onDelete())
      }

      expect(onMockSelected.mock.calls[0][0]).toEqual(['Tsp'])
    })
  })
})
