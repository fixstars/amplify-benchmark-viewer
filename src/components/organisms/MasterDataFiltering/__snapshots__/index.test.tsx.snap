// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MasterDataFiltering /> Rendered well 1`] = `
<div>
  <div
    class="MuiBox-root css-1itv5e3"
  >
    <div
      class="MuiBox-root css-1yp4ln"
    >
      <div
        class="MuiBox-root css-1yosmn2"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          Problem Class Filtering
        </p>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
            data-testid="AddCircleIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiBox-root css-0"
      />
    </div>
  </div>
</div>
`;

exports[`<MasterDataFiltering /> Rendered well with selected data 1`] = `
<div>
  <div
    class="MuiBox-root css-1itv5e3"
  >
    <div
      class="MuiBox-root css-1yp4ln"
    >
      <div
        class="MuiBox-root css-1yosmn2"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          Problem Class Filtering
        </p>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
            data-testid="AddCircleIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiBox-root css-0"
      >
        <div
          class="MuiButtonBase-root MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorPrimary MuiChip-deletable MuiChip-deletableColorPrimary MuiChip-outlinedPrimary css-4ikpb8-MuiButtonBase-root-MuiChip-root"
          role="button"
          tabindex="0"
        >
          <span
            class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
          >
            Tsp
          </span>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-deleteIcon MuiChip-deleteIconSmall MuiChip-deleteIconColorPrimary MuiChip-deleteIconOutlinedColorPrimary css-i4bv87-MuiSvgIcon-root"
            data-testid="CancelIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
            />
          </svg>
        </div>
        <div
          class="MuiButtonBase-root MuiChip-root MuiChip-outlined MuiChip-sizeSmall MuiChip-colorPrimary MuiChip-deletable MuiChip-deletableColorPrimary MuiChip-outlinedPrimary css-4ikpb8-MuiButtonBase-root-MuiChip-root"
          role="button"
          tabindex="0"
        >
          <span
            class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
          >
            Sudoku
          </span>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-deleteIcon MuiChip-deleteIconSmall MuiChip-deleteIconColorPrimary MuiChip-deleteIconOutlinedColorPrimary css-i4bv87-MuiSvgIcon-root"
            data-testid="CancelIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
